### ISSUE TITLE
syntax: consider zsh support

### ISSUE THREAD

**Issue by caarlos0** [mvdan/sh#120]
Congratz for the great tool!

Would be nice to have zsh support.

**Comment by mvdan** [comment]
Zsh is a massively complex shell - even more so than Bash - so let me ask a few questions first.

Why? Is there much Zsh code in the wild? From what I can tell, Bash, POSIX Shell and mksh already cover pretty much everything.

Is there a language spec for Zsh? If there's none, is there a document that summarizes all the language features?

The bottom line is that I fear this would be an insane amount of work for little benefit. I never use Zsh, so there's also that - I have no personal incentive to do it.

Also /cc @D630 who suggested zsh alongside mksh (the last one I did, it wasn't much work on top of Bash/POSIX)

**Comment by mvdan** [comment]
Also note that I would put Zsh in a similar spot as Fish. They're cooler, newer shells that - as far as I know - are popular as interactive shells. But not so much as in shell scripts that you maintain over time. Those tend to be in Bash or POSIX Shell. This is the kind of shell code that `shfmt` is for.

**Comment by caarlos0** [comment]
> Also note that I would put Zsh in a similar spot as Fish. They're cooler, newer shells that - as far as I know - are popular as interactive shells. But not so much for scripts that you maintain over time. Those tend to be in Bash or POSIX Shell, since you want the scripts to run on as many systems as possible. This is the kind of shell code that shfmt is for.

yeah, I agree.

**Comment by incognitoRepo** [comment]
hello, no disagreements here. but I saw a very similar response in the equally "dam this is useful" shellcheck repo (im coming from vscode). now, im no shell expert, but i always had the impression zsh was the #2 game in town. at least, anyone who seems to care about aesthetics is almost guaranteed to be on zsh. its actually a bit disappointing to learn that its not that important (i.e., major tools don't support it). and really, i guess thats more on zsh people (who could make a spec or a pull req or smth). anyhow, thats just for context. great package!

**Comment by mvdan** [comment]
@incognitoRepo please note my earlier point about interactive shells. zsh may be popular as an interactive shell, but this tool is for scripts and programs one maintains. Can anyone provide proof that zsh is popular as a scripting language too? I've personally never come across a script in it, but I might be biased.

That's just as far as utility goes; after that, we'd also have to consider how much work it would be. I probably wouldn't have the time to impement all of its syntax features, particularly since bash support isn't even complete yet :)

**Comment by texastoland** [comment]
As of macOS Catalina (the current version) it's the default shell for new user accounts. It may result in more requests but I agree it won't increase usage in real projects for a long time.

> Is there a language spec for Zsh? If there's none, is there a document that summarizes all the language features?

There's a pretty hefty [manual](http://zsh.sourceforge.net/Doc/Release/zsh_toc.html#SEC_Contents) ([PDF](http://zsh.sourceforge.net/Doc/zsh_a4.pdf)) that I mainly reference when I see a glob or expansion I'm not familiar with. I didn't see any formal grammar either.

**Comment by mvdan** [comment]
Huh, interesting, I didn't know Mac had switched over. I'm not sure if that will result in people writing more zsh scripts, though.

**Comment by mvdan** [comment]
Different people have pinged me about the macOS move in the past few weeks, so let's reopen this issue for now.

Pros:

* Supporting sh, bash, and mksh, it feels like zsh is the next logical step by popularity
* Being the default shell of new macOS versions should mean that macOS users would start writing some zsh instead of bash
* zsh is pretty similar to bash/mksh, so syntax-wise, we should be able to reuse a good amount of code

Cons:

* Like bash, zsh is a very complex shell. bash work started nearly three years ago, and even today I still find bugs or missing features in the syntax package.
* Like bash and mksh, zsh does not have a formal language specification. This will make it harder to map out what language features we're missing.
* Increase in complexity, which would hurt binary sizes, compile times, and my ability to maintain this project long-term.

I'm honestly about 50/50 on this right now. Thoughts welcome; please keep it concise so that the thread doesn't derail.

**Comment by theclapp** [comment]
I used to use zsh daily.  (More and more now I use mvdan/sh/interp! :)  I wrote some scripts in it, almost entirely for my own use.  My workplace(s) for years have generally been bash shops, so anything I wanted to share, I felt like I had to either heavily comment / document it, or just translate to bash.  (In fairness, that wasn't usually a huge burden, but it was usually non-zero, nevertheless.)

So I like zsh, don't get me wrong!

But I also agree that it's very large, and kind of ad-hoc.  It does kind of feel like there's an underlying logic to it, but that can be hard to suss out / reverse engineer.

I think it'd be fair to say "zsh maintainer wanted", and maybe implement often-used, low-hanging fruit, as time, demand, and contributions allow.  (I was going to mention `**` as a feature in zsh I miss a lot in bash, but I checked the bash manual and it turns out bash can do that, so I guess the joke's on me there.  :)

Have you considered inquiring on the zsh mailing lists for collaborators?

**Comment by mvdan** [comment]
> But I also agree that it's very large, and kind of ad-hoc.

The good news here is that I'm only considering parser and printer support, i.e. just the syntax. Zsh is different from Bash in lots of ways when it comes to running a script, but we don't care because I don't think it would ever be a good idea to add a "zsh mode" in the interpreter.

> implement often-used, low-hanging fruit, as time, demand, and contributions allow.

This would definitely be a best-effort basis :) That's how we're doing Bash too. It would be insane to try to reach 100% syntax compatibility in a single release.

> Have you considered inquiring on the zsh mailing lists for collaborators?

That's a good idea. I guess I can phrase it as an FYI, to see if there would be interest from any zsh developers or heavy users.

**Comment by texastoland** [comment]
I keep using shfmt and shellcheck with my zsh files. Most of the differences are semantic rather than syntactic. Here's an obtuse yet idiomatic example that doesn't parse:

```shell
#                ┌ nesting
#                │  ┌ flags ┌ modifiers
#                ▼  ▼       ▼
export ZDOTDIR=${${(%):-%x}:P:h}
# ...globbing also has additions
```

I see three potential levels of support:

1. Encourage encapsulation of unparseable code in separate files. It hasn't been too inconvenient for personal scripts.
2. Support some comment to skip parsing the next line.
3. Only support differences at the highest level. My example highlights the additions to expansions. I guess there could be corner cases but you could test against one of the frameworks for the majority case.

**Comment by mvdan** [comment]
> Encourage encapsulation of unparseable code in separate files.

That's pretty much what we have today though, no?

> Support some comment to skip parsing the next line.

I don't think we want to do this. The parser should just parse code statically, without understanding special comments or anything like that.

> Only support differences at the highest level.

I'm not sure if I understand. If we support zsh syntax, we should aim at supporting all of its documented syntax, just like we do with bash.

**Comment by texastoland** [comment]
> That's pretty much what we have today though, no?

That's what I meant.

> I don't think we want to do this. 

To me it's the simplest and friendliest solution in the meantime.

> The parser should just parse code statically, without understanding special comments or anything like that.

I didn't understand about statically. It's essentially like a comment but a block ending after the first non-comment line instead of some token.

> If we support zsh syntax, we should aim at supporting all of its documented syntax, just like we do with bash.

For instance expansions can start with flags in parentheses (`${(%)...}` in my example). Those flags can take arguments themselves separated by potentially arbitrary characters similar to pattern expansions. Rather than fully suss it out you could leave everything in those parentheses like a "todo" in the AST.

**Comment by texastoland** [comment]
I thought the only syntactic differences were related to expansion and globbing but I recently stumbled on code in the wild using [alternate syntax for control structures](http://zsh.sourceforge.net/Doc/Release/Shell-Grammar.html#Alternate-Forms-For-Complex-Commands).

**Comment by texastoland** [comment]
> * Support some comment to skip parsing the next line.

I ended up implementing this with a wrapper script:

```bash
# match: # shfmt-no-parse: explain here
local pattern='^\s*#\s*shfmt-no-parse.*\s+'
# comment next lines
fastmod --accept-all $pattern '$0#' "$@"
# error: shfmt -ws
shfmt -w -s "$@"
# uncomment them again
fastmod --accept-all "(${pattern})#" '$1' "$@"
```

**Comment by ajeetdsouza** [comment]
While there may not be as many small zsh scripts in the wild, there are definitely far more _plugins_, which are large codebases written in pure zsh. I think they would stand to benefit the most from a formatter. Some of the ones I use:

- [powerlevel10k](https://github.com/romkatv/powerlevel10k) - 20k+ lines of zsh
- [fast-syntax-highlighting](https://github.com/zdharma/fast-syntax-highlighting) - 1.5k+ lines of zsh
- [ohmyzsh](https://github.com/ohmyzsh/ohmyzsh) framework - 20k+ lines of zsh

I already use shfmt for [zoxide](https://github.com/ajeetdsouza/zoxide)'s bash and posix plugins, I would love to use it for zsh as well.

**Comment by mvdan** [comment]
I'm strongly leaning towards implementing zsh support in the syntax package (and shfmt). The only question is allocating the 2-3 weeks of continuous work to get to a working prototype :)

If you would like to help by testing in the future, please react with the "eyes" emoji on this comment and I'll be in touch.

**Comment by docwhat** [comment]
IIRC zsh itself can dump out an abstract syntax tree.  That could be useful for testing.

**Comment by mvdan** [comment]
@docwhat very interesting - can you share how to do that?

**Comment by docwhat** [comment]
Ah, here we go...

In `zshexpn` under [`Parameter Expansion Flags`](https://zsh.sourceforge.io/Doc/Release/Expansion.html#Parameter-Expansion-Flags):

> Split the result of the expansion into words using shell parsing to find the words, i.e. taking into account any quoting in the value. Comments are not treated specially but as ordinary strings, similar to interactive shells with the `INTERACTIVE_COMMENTS` option unset (however, see the `Z` flag below for related options)
>
> Note that this is done very late, even later than the `(s)` flag. So to access single words in the result use nested expansions as in `${${(z)foo}[2]}`. Likewise, to remove the quotes in the resulting words use `${(Q)${(z)foo}}`.

### Example

```zsh
#!/bin/zsh

read -r -d '' codetext <<'CODE'
# A function to help with diction.
the_rain() {
  local rain="falls mainly in the plain"
  echo "in spain ${(qq)rain}"
}
# end of the_rain
CODE

declare -r -a codearr=( "${(Z:c:@)codetext}" )

declare -i c=0 i=0
declare last=';'
for item in "${(@)codearr}"; do
  [[ $item == '{' ]] && ((i++))
  [[ $item == '}' ]] && ((i--))
  if [[ $last == ';' ]]; then
    for i in {0..$i}; do echo -n '  '; done
  fi
  if [[ $item == ';' ]]; then
    echo
  else
    tput setaf $((c + 1))
    echo -n ">${item}< "
    c=$(( ((c+1) % 7)))
  fi
  last=$item
done

echo
```

If you run it you get this (except the colors don't show up here):

```
  ># A function to help with diction.< 
  >the_rain< >()< >{< 
    >local< >rain="falls mainly in the plain"< 
    >echo< >"in spain ${(qq)rain}"< 
  >}< 
  ># end of the_rain< 
```

**Comment by marlonrichert** [comment]
@docwhat @mvdan 

There's no need to use Zsh to parse the code; Zsh can actually _format_ it for you. 

If you put the code into a function, then `type -f` can be used to print out fully formatted code, minus any comments or blank lines. For example:
```zsh
% setopt interactivecomments
% tmp() {
#!/bin/zsh
read -r -d '' codetext <<'CODE'
# A function to help with diction.
the_rain() {
  local rain="falls mainly in the plain"
  echo "in spain ${(qq)rain}"
}
# end of the_rain
CODE
declare -r -a codearr=( "${(Z:c:@)codetext}" ); declare -i c=0 i=0; declare last=';'
for item in "${(@)codearr}"; do
[[ $item == '{' ]] && ((i++)); [[ $item == '}' ]] && ((i--))
if [[ $last == ';' ]]; then
for i in {0..$i}; do echo -n '  '; done
fi
if [[ $item == ';' ]]; then echo; else
tput setaf $((c + 1))
echo -n ">${item}< "
c=$(( ((c+1) % 7)))
fi
last=$item
done
echo
}
% type -f tmp 
tmp () {
	read -r -d '' codetext <<'CODE'
# A function to help with diction.
the_rain() {
  local rain="falls mainly in the plain"
  echo "in spain ${(qq)rain}"
}
# end of the_rain
CODE
	declare -r -a codearr=("${(Z:c:@)codetext}") 
	declare -i c=0 i=0 
	declare last=';' 
	for item in "${(@)codearr}"
	do
		[[ $item == '{' ]] && ((i++))
		[[ $item == '}' ]] && ((i--))
		if [[ $last == ';' ]]
		then
			for i in {0..$i}
			do
				echo -n '  '
			done
		fi
		if [[ $item == ';' ]]
		then
			echo
		else
			tput setaf $((c + 1))
			echo -n ">${item}< "
			c=$(( ((c+1) % 7))) 
		fi
		last=$item 
	done
	echo
}
%
```
All that's left to do after that is figure out where to reinsert the comments and blank lines.

**Comment by denysdovhan** [comment]
Hey, I'm the author of @spaceship-prompts (almost 17K stars and 12K lines of code). Looking forward to using `shfmt` against my codebase 🙂

**Comment by yutkat** [comment]
I found this one.
https://github.com/lovesegfault/beautysh

**Comment by marlonrichert** [comment]
@yutkat Doesn't look like that has Zsh support either, though.

**Comment by yutkat** [comment]
Although support is not explicitly noted, but it worked fine in my zsh program.

**Comment by 0xdevalias** [comment]
Currently the main/only thing i've noticed `shfmt` breaking within my zsh scripts is this sort of syntax:

```diff
-if (( $+commands[git] ))
-then
+if (($ + commands[git])); then
```

Where the `$ + command` (instead of `$+command`) then becomes invalid zsh script, and causes the script to break.

**Comment by pboushy** [comment]
I'm a "MacAdmin" working at a large company.
Our rule is if you can write the script easily in POSIX sh, do that.
If your script need arrays or anything complex use ZSH.

I know many other companies are doing the same.

ScriptingOSX.com is one of the main websites for learning how to script on macOS, and all his latest stuff is ZSH.

Adding ZSH support would be awesome.

I mentioned to the shellcheck maintainer that I'd be willing to contribute if they can provide some guidance on bite-size pieces to tackle first. But then I discovered it's written in Haskell... which requires learning an entirely new language for me on top of taking on a complex enough area already. While I don't have a lot of experience with go, I'm not brand new to it.

**Comment by rseichter** [comment]
Here's an example of a valid ZSH statement which is not valid for BASH. The statement is part of my .zshrc and works fine there.

```
. "$XDG_CACHE_HOME/p10k-instant-prompt-${(%):-%n}.zsh"
```

`shfmt -w filename` returns the error "parameter expansion requires a literal". Out of curiosity, I also tried `shellcheck -s bash filename`, resulting in [error SC2296](https://www.shellcheck.net/wiki/SC2296).

I use ZSH on all my macOS and Linux based machines, and I would appreciate ZSH support in `shfmt`. Thanks.

**Comment by jansorg** [comment]
After implementing initial Zsh support for [BashSupport Pro](https://www.bashsupport.com) I can confirm that it would be a huge amount of work to implement Zsh support for shfmt.
Most notable differences I've found:
- [Parameter expansion flags](https://www.bashsupport.com/zsh/parameter-expansion-flags/), e.g. `${(aO)values}` or even with additional arguments `${(g:opts:)name}`
- [Modifiers](https://www.bashsupport.com/zsh/modifiers/), e.g. `$relativePath:a` which are used for globs and variables.
- [Alternate forms](https://zsh.sourceforge.io/Doc/Release/Shell-Grammar.html#Alternate-Forms-For-Complex-Commands) of complex commands, e.g. `while list { list }`
- Subtle differences in syntax support, e.g. command separators in blocks are optional `{ myCommand }` instead of Bash's `{ myCommand; }`
- Extensions like multiple names passed to function definitions `function a b c { body; }`, anonymous functions like `function { local var="only-here"; echo $@;} a b c `
- [Extended arithmetic operators](https://zsh.sourceforge.io/Doc/Release/Arithmetic-Evaluation.html), e.g. `&&=`
- The ["Zsh amper bang"](https://stackoverflow.com/questions/20473254/whats-the-difference-between-and-in-zsh), e.g. `a &| b`
- and a bunch more

tl;dr Zsh is an even more complex language than Bash and adding support would take a lot of time

**Comment by mvdan** [comment]
Fully agreed with @jansorg, even more so because both bash and zsh are languages that work "as documented and implemented" without a formal spec, so they tend to have undocumented behaviors and evolve rapidly to add new features. At the same time, `shfmt` can work OK for many people in practice with partial support for the most commonlly used features. Just look at the earliest releases of this software - bash support was very limited and buggy, but some people still used it.

I also wanted to give signs of life, as I haven't [given an update in over two years](https://github.com/mvdan/sh/issues/120#issuecomment-903077180). This is very much still planned, but I'll keep quiet until I find the time to work on a prototype, because it doesn't feel right to make promises on ETAs.

**Comment by rseichter** [comment]
I am certain that adding ZSH support is not an easy task, and my previous comment was mostly meant to indicate that I am a member of the potentially growing group of people who would be grateful if shfmt supported the ZSH language. No pressure, just something on my wish list worth mentioning. 😉

**Comment by mvdan** [comment]
No worries, this thread and the added comments aren't causing me any sort of stress right now. I just noticed it's already been over two years since I last gave an update (geez, already?) so I wanted to briefly update again.

**Comment by thesobercoder** [comment]
I gave up hope and started using Copilot to do the formatting for me.

**Comment by telemachus** [comment]
@stephanschielke At a glance, beautysh only mentions bash. Do you have reason to think that it supports zsh better than this repo? (I don't think that it should since it seems more limited in scope.)

**Comment by mvdan** [comment]
Two more features are done - https://github.com/mvdan/sh/issues/1200 and https://github.com/mvdan/sh/issues/1201.

I have gone through @jansorg's updated comment once more, and did a bit of googling, and populated https://github.com/mvdan/sh/labels/zsh with six more issues. I believe the list of eight issues in total is a good first milestone for Zsh support to be good enough for early testing. I don't think any of these should take a significant amount of time either, from a quick look - given that we just need to add parsing and printing support, but not interpreting.

If I have missed any popular Zsh syntax feature that isn't already done at master or captured in that list, please open a new issue. Thanks!

**Comment by costincaraivan** [comment]
> Also note that I would put Zsh in a similar spot as Fish. They're cooler, newer shells that - as far as I know - are popular as interactive shells. But not so much for scripts that you maintain over time. Those tend to be in Bash or POSIX Shell, since you want the scripts to run on as many systems as possible. This is the kind of shell code that shfmt is for.

Thank you for the awesome tool. I'm keeping an eye on zsh support 🤞 

Very small nitpick about the "cooler, newer shells" 🙂

Bash - Initial release 8 June 1989; 36 years ago
Zsh - Initial release 1990; 35 years ago

I was using zsh 15 years ago, it's not a new shell. Bash was the GNU shell and GNU was a big deal in the 80s and 90s, so that's why it became a lot more popular early on.

**Comment by tmccombs** [comment]
> popular as interactive shells. But not so much for scripts that you maintain over time.

Zsh (and fish) are often used for fairly large projects maintained over time for "plugins" for  interactive use. 

That said, adding zsh support would no doubt be a large undertaking.

**Comment by mvdan** [comment]
`$foo[1]` is done, as are #1207, #1228, and #1205. I'll probably release v3.13.0 with the added features and API as of master in a few days, so please give it a try and let me know of any bugs or issues you spot with it.

Zsh support is not complete - there are four issues left open. Nor is Bash support complete, to be clear. But I feel like this is a decent point for a more general announcement, and we're overdue for a release.

**Comment by mvdan** [comment]
Thanks to all who have been reporting bugs, especially @LangLangBart. I have been prioritizing fixing those which seemed like regressions or needed API changes which would be hard to do once a release is already done.

So, a little bit later than I hoped, but now I do think we're at a point where I can release. Some issues still remain, but those can be done later and incrementally.

**Comment by mvdan** [comment]
With v3.13.1 out and the amount of incoming zsh bug reports slowing down, I think we can call the initial zsh support done :) Please upvote or raise issues that affect you and I'll look at them. We probably won't get to 100% compatibility, but we never got there with Bash and in practice that's been fine.

**Comment by mdekstrand** [comment]
@mvdan thank you so much! I've been using it successfully, and 3.13.1 is handling the trickiest things that were causing me problems in earlier versions.

**Comment by mvdan** [comment]
Good to hear :) I should also mention that anyone benefitting from this work, especially companies, can sponsor me on GitHub as well. I recently added tiers with benefits, both for recurring and one-time contributions.

**Comment by carhartl** [comment]
Thank you, thank you, thank you! I loved when this was added..

**Comment by ch3rub1m** [comment]
I never imagined I'd live to see zsh support completed; it's amazing!

**Comment by HaleTom** [comment]
Linking this as it's quite related:
- https://ewhauser.github.io/shuck/
- https://github.com/ewhauser/shuck

40x+ speedup compared with `shellcheck`.

**Comment by mvdan** [comment]
I'm sorry @HaleTom, but how does that relate to this project or issue in any way?

**Comment by felipecrs** [comment]
Perhaps because it claims to be a formatter while as well supporting zsh.
